src/persist/src/operators/stream.rs

```diff
@@ -739,114 +742,27 @@ where
-                            continue;
-                        }
-                        session.give(update);
-                    }
-
-                    if !to_retract.is_empty() {
-                        let write_future = write.write(&to_retract);
-                        to_retract.clear();
-
-                        // We are not using the capability for the main output later, but we are
-                        // holding on to it to keep the frontier from advancing because that frontier
-                        // is used downstream to track how far we have persisted. This is used, for
-                        // example, by upsert()/persist()/seal()/conditional_seal() operators
-                        // and await_frontier().
-                        pending_futures.push_back((
-                            cap.delayed(cap.time()),
-                            cap.retain_for_output(error_output_port),
-                            write_future,
-                        ));
-                    }
-                });
-
-                // Swing through all pending futures and see if they're ready. Ready futures will
-                // invoke the Activator, which will make sure that we arrive here, even when there
-                // are no changes in the input frontier or new input.
-                let waker = futures_util::task::waker_ref(&activator);
-                let mut context = Context::from_waker(&waker);
-
-                while let Some((cap, error_cap, pending_future)) = pending_futures.front_mut() {
-                    match Pin::new(pending_future).poll(&mut context) {
-                        std::task::Poll::Ready(result) => {
-                            match result {
-                                Ok(seq_no) => {
-                                    log::trace!(
-                                        "In {}, finished writing for time: {}, seq_no: {:?}",
-                                        &operator_name,
-                                        cap.time(),
-                                        seq_no,
-                                    );
-                                }
-                                Err(e) => {
-                                    let mut session = error_output.session(&error_cap);
-                                    let error = format!(
-                                        "In {}, error writing data for time {}: {}",
-                                        &operator_name,
-                                        error_cap.time(),
-                                        e
-                                    );
-                                    log::error!("{}", error);
+        let (pass_through, to_retract) =
+            self.branch(move |_, (_, t, _)| if t < &upper_ts { false } else { true });
+
+        let (retract_oks, errs) = to_retract
+            .map(|(data, time, diff)| (data, time, -diff))
+            .persist(name, write);
```

Comment:
It would be good if we could encode the "retract" nature of these in the name. So that we continue to have useful logging and operator names.